### PR TITLE
feat(coderd): add `times_used` to `coder_app`s in insights API

### PR DESCRIPTION
**Pull request by mafredri**
For now, only applied to `coder_app`s, same logic can be implemented for
VS Code, SSH, etc.

Part of #13099

### CODE REVIEW

**Comment by mafredri** [comment]
* **#13292** <a href="https://app.graphite.dev/github/pr/coder/coder/13292?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mafredri and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/database/queries/insights.sql [hunk 1]
```diff
@@ -249,7 +249,7 @@ WITH
 			apps.slug,
 			apps.display_name,
 			apps.icon,
-			tus.app_usage_mins
+			(tus.app_usage_mins -> apps.slug)::smallint AS usage_mins
```
Comment (mafredri) [inline coderd/database/queries/insights.sql:252]: Drive-by optimization to avoid `jsonb_each` in the following CTE.

File: coderd/database/dbmem/dbmem.go [hunk 2]
```diff
@@ -3149,6 +3149,30 @@ func (q *FakeQuerier) GetTemplateAppInsights(ctx context.Context, arg database.G
...
+				display_name,
+				icon,
+				-- Turn start_time into a unique identifier that identifies a users
+				-- continuous app usage. The value of uniq is otherwise garbage.
+				--
+				-- Since we're aggregating per user app usage across templates,
+				-- there can be duplicate start_times. To handle this, we use the
+				-- dense_rank() function, otherwise row_number() would suffice.
+				start_time - (
+					dense_rank() OVER (
```
Comment (mtojek) [inline coderd/database/dbmem/dbmem.go:3166]: TIL `dense_rank()`

File: coderd/database/queries/insights.sql [hunk 3]
```diff
@@ -273,14 +273,36 @@ WITH
...
+			template_usage_stats_with_apps
 		GROUP BY
 			start_time, user_id, slug, display_name, icon
 	),
+	-- Analyze the users unique app usage across all templates. Count
+	-- usage across consecutive intervals as continuous usage.
+	times_used AS (
+		SELECT DISTINCT ON (user_id, slug, display_name, icon, uniq)
+			slug,
+			display_name,
```
Comment (mtojek) [inline coderd/database/queries/insights.sql:287]: QQ: If somebody changes the `display_name` for the app, will it break the calculation? Probably a question related with other subqueries too.

File: coderd/database/queries/insights.sql [hunk 4]
```diff
@@ -302,7 +324,17 @@ SELECT
 	ai.slug,
 	ai.display_name,
 	ai.icon,
-	(SUM(ai.usage_mins) * 60)::bigint AS usage_seconds
+	(SUM(ai.usage_mins) * 60)::bigint AS usage_seconds,
+	COALESCE((
+		SELECT
+			COUNT(*)
```
Comment (mtojek) [inline coderd/database/queries/insights.sql:330]: It is just about cardinality of app usage, right?

File: coderd/database/queries/insights.sql [hunk 5]
```diff
@@ -273,14 +273,36 @@ WITH
...
+			start_time - (
+				dense_rank() OVER (
+					PARTITION BY
+						user_id, slug, display_name, icon
+					ORDER BY
+						start_time
+				) * '30 minutes'::interval
+			) AS uniq
+		FROM
+			template_usage_stats_with_apps
```
Comment (johnstcn) [inline coderd/database/queries/insights.sql:304]: What sort of additional load does this place on the DB?

**Review by johnstcn — APPROVED** [review]
LGTM, but would be interesting to see an EXPLAIN before vs an EXPLAIN after.

File: coderd/database/queries/insights.sql (hunk 3 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:287]: It'll handle it according to current behavior -- a change in `display_name` will be treated as a new app.

File: coderd/database/queries/insights.sql (hunk 5 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:304]: I didn't measure but I don't expect it to be a lot, this is similar in weight (or lighter) as the `app_insights` CTE and could theoretically be run in parallel (don't know if PG will, though).

File: coderd/database/queries/insights.sql (hunk 4 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:330]: Yes, we are only interested in rows of this table, not of it's content.

**Comment by mafredri** [comment]
> LGTM, but would be interesting to see an EXPLAIN before vs an EXPLAIN after.

Don't have a good dataset to test, but with a little bit of data it looks like this:

- Before: https://explain.dalibo.com/plan/2f47f5h3c84145g7
- After: https://explain.dalibo.com/plan/6h5e82gb7b8c79dc

File: coderd/database/queries/insights.sql (hunk 5 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:304]: I posted the explains, it seems to be parallel and very cheap.

**Review by mtojek — APPROVED** [review]
👍

File: coderd/database/queries/insights.sql (hunk 3 above)
Comment (matifali) [reply coderd/database/queries/insights.sql:287]: But it should nit as technically the app is the same app.

Comment (mafredri) [reply coderd/database/queries/insights.sql:287]: It depends, in aggregation across the whole deployment (which we don't show, but is possible), the apps might not be the same at all. For now I would not want to introduce one feature that behaves differently from the rest, we already show multiple entries for apps and parameters if something (like the display name) has changed.

Comment (matifali) [reply coderd/database/queries/insights.sql:287]: Yes. I am aware if the existing behavior and suggesting that we should change it sometimes in the future to be more robust to cosmetic changes to a `coder_app`.

Comment (mafredri) [reply coderd/database/queries/insights.sql:287]: @matifali actually, have you recently confirmed this is still a problem? I think this is mainly a problem with parameters. If the app name changes it should only use the most recent name, at least since the rollup refactor.

**Comment by stirby** [comment]
Awesome!
